File: src/test/java/org/reflections/ReflectionsThreadSafenessTest.java
package org.reflections;

import com.google.common.collect.ImmutableMap;
import org.junit.Test;
import org.reflections.scanners.SubTypesScanner;
import org.reflections.util.ClasspathHelper;
import org.reflections.util.ConfigurationBuilder;

import java.util.Set;
import java.util.concurrent.*;

import static java.util.Collections.singletonList;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.junit.Assert.assertEquals;

public class ReflectionsThreadSafenessTest {

    /**
     * https://github.com/ronmamo/reflections/issues/81
     */
    @Test
    public void reflections_scan_is_thread_safe() throws Exception {

        Callable<Set<Class<? extends ImmutableMap>>> callable = () -> {
            Reflections reflections = new Reflections(new ConfigurationBuilder().setUrls(singletonList(ClasspathHelper.forClass(
                    ImmutableMap.class))).setScanners(new SubTypesScanner(false)));

            return reflections.getSubTypesOf(ImmutableMap.class);
        };

        ExecutorService pool = Executors.newFixedThreadPool(2);

        Future<?> first  = pool.submit(callable);
        Future<?> second = pool.submit(callable);

        assertEquals(first.get(5, SECONDS), second.get(5, SECONDS));
    }
}

File: src/test/java/org/reflections/ReflectionsThreadSafenessTest.kt
package org.reflections

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit.SECONDS

class ReflectionsThreadSafenessTest {
    /**
     * https://github.com/ronmamo/reflections/issues/81
     */
    @Test
    fun reflections_scan_is_thread_safe() {
        val callable = {
            SubTypesScanner(excludeObjectClass = false)
                .scan(Map::class.java).subTypesOf(Map::class.java)
        }

        val pool = Executors.newFixedThreadPool(2)

        val first = pool.submit(callable)
        val second = pool.submit(callable)

        assertEquals(first.get(10, SECONDS), second.get(10, SECONDS))
    }
}
